redfish/provider/resource_redfish_storage_volume.go

```diff
@@ -1,5 +1,757 @@
+		return diags
+	}
+	drives, err := getDrives(allStorageDrives, driveNames)
+	if err != nil {
+		diags.AddError("Error when getting the drives", err.Error())
+		return diags
+	}
+
+	newVolume := map[string]interface{}{
+		"VolumeType":         volumeType,
+		"DisplayName":        volumeName,
+		"Name":               volumeName,
+		"ReadCachePolicy":    readCachePolicy,
+		"WriteCachePolicy":   writeCachePolicy,
+		"CapacityBytes":      capacityBytes,
+		"OptimumIOSizeBytes": optimumIOSizeBytes,
+		"Oem": map[string]map[string]map[string]interface{}{
+			"Dell": {
+				"DellVolume": {
+					"DiskCachePolicy": diskCachePolicy,
+				},
+			},
+		},
+		"@Redfish.OperationApplyTime": applyTime,
+	}
+
+	var listDrives []map[string]string
+	for _, drive := range drives {
+		storageDrive := make(map[string]string)
+		storageDrive["@odata.id"] = drive.Entity.ODataID
+		listDrives = append(listDrives, storageDrive)
+	}
+	newVolume["Drives"] = listDrives
+
+	// Create volume job
+	jobID, err := createVolume(service, storage.ODataID, newVolume)
+	if err != nil {
+		diags.AddError("Error when creating the virtual disk on disk controller", err.Error())
+		return diags
+	}
+
+	// Immediate or OnReset scenarios
+	if applyTime == string(redfishcommon.OnResetApplyTime) { // OnReset case
+		// Get reset_timeout and reset_type from schema
+		resetType := d.ResetType.ValueString()
+		resetTimeout := d.ResetTimeout.ValueInt64()
+
+		// Reboot the server
+		pOp := powerOperator{ctx, service}
+		_, err := pOp.PowerOperation(resetType, resetTimeout, intervalStorageVolumeJobCheckTime)
+		if err != nil {
+			diags.AddError(RedfishJobErrorMsg, err.Error())
+			return diags
+		}
+	}
+
+	// Wait for the job to finish
+	err = common.WaitForJobToFinish(service, jobID, intervalStorageVolumeJobCheckTime, volumeJobTimeout)
+	if err != nil {
+		diags.AddError(RedfishJobErrorMsg, err.Error())
+		return diags
+	}
+
+	// Get storage volumes
+	volumes, err := storage.Volumes()
+	if err != nil {
+		diags.AddError("there was an issue when retrieving volumes", err.Error())
+		return diags
+	}
+	volumeID, err := getVolumeID(volumes, volumeName)
+	if err != nil {
+		diags.AddError("Error. The volume ID with volume name %s on %s controller was not found", err.Error())
+		return diags
+	}
+
+	d.ID = types.StringValue(volumeID)
+	diags = readRedfishStorageVolume(service, d)
+	return diags
+}
+
+func readRedfishStorageVolume(service *gofish.Service, d *models.RedfishStorageVolume) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	// Check if the volume exists
+	_, err := redfish.GetVolume(service.GetClient(), d.ID.ValueString())
+	if err != nil {
+		e, ok := err.(*redfishcommon.Error)
+		if !ok {
+			diags.AddError("There was an error with the API", err.Error())
+			return diags
+		}
+		if e.HTTPReturnedStatusCode == http.StatusNotFound {
+			log.Printf("Volume %s doesn't exist", d.ID.ValueString())
+			d.ID = types.StringValue("")
```

Comment:
I dont think this is doing the intended thing. I think we are supposed to recreate the volume if it has been somehow deleted.
One way to get it to work wd be to remove the state.

Reply:
changed to remove resource if volume doesn't exist.